ipn/ipnlocal/web.go

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build !ios
+
+package ipnlocal
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"tailscale.com/client/tailscale"
+	"tailscale.com/client/web"
+	"tailscale.com/envknob"
+)
+
+// webServer holds state for the web interface for managing
+// this tailscale instance. The web interface is not used by
+// default, but initialized by calling LocalBackend.WebOrInit.
+type webServer struct {
+	ws         *web.Server  // or nil, initialized lazily
+	httpServer *http.Server // or nil, initialized lazily
+
+	// lc optionally specifies a LocalClient to use to connect
+	// to the localapi for this tailscaled instance.
+	// If nil, a default is used.
+	lc *tailscale.LocalClient
+
+	wg sync.WaitGroup
+}
+
+// SetWebLocalClient sets the b.web.lc function.
+// If lc is provided as nil, b.web.lc is cleared out.
+func (b *LocalBackend) SetWebLocalClient(lc *tailscale.LocalClient) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.web.lc = lc
+}
+
+// WebInit initializes the web interface for managing
+// this tailscaled instance. If the web interface is
+// already running, WebInit is a no-op.
+func (b *LocalBackend) WebInit() (err error) {
+	if !envknob.Bool("TS_DEBUG_WEB_UI") {
+		return errors.New("web ui flag unset")
+	}
+
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if b.web.ws != nil {
+		return nil
+	}
+
+	b.logf("WebInit: initializing web ui")
+	if b.web.ws, err = web.NewServer(web.ServerOpts{
+		// TODO(sonia): allow passing back dev mode flag
+		LocalClient: b.web.lc,
+		Logf:        b.logf,
+	}); err != nil {
+		return fmt.Errorf("web.NewServer: %w", err)
+	}
+
+	// Start up the server.
+	b.web.wg.Add(1)
+	go func() {
+		defer b.web.wg.Done()
+		// TODO(sonia/will): only listen on Tailscale IP addresses
+		addr := ":5252"
+		b.web.httpServer = &http.Server{
+			Addr:    addr,
+			Handler: http.HandlerFunc(b.web.ws.ServeHTTP),
+		}
+		b.logf("WebInit: serving web ui on %s", addr)
+		if err := b.web.httpServer.ListenAndServe(); err != nil {
+			if err != http.ErrServerClosed {
+				b.logf("[unexpected] WebInit: %v", err)
+			}
+		}
+	}()
+
+	b.logf("WebInit: started web ui")
+	return nil
+}
+
+// WebShutdown shuts down any running b.web servers and
+// clears out b.web state (besides the b.web.lc field,
+// which is left untouched because required for future
+// web startups).
+func (b *LocalBackend) WebShutdown() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.webShutdownLocked()
+}
+
+// webShutdownLocked shuts down any running b.web servers
+// and clears out b.web state (besides the b.web.lc field,
+// which is left untouched because required for future web
+// startups).
+//
+// b.mu must be held.
+func (b *LocalBackend) webShutdownLocked() {
+	if b.web.ws != nil {
+		b.web.ws.Shutdown()
+	}
+	if b.web.httpServer != nil {
+		if err := b.web.httpServer.Shutdown(context.Background()); err != nil {
+			b.logf("[unexpected] webShutdownLocked: %v", err)
+		}
```

Comment:
could this block for a long time?

Reply:
Hmm, good question. Yeah I guess theoretically it could. We could pass a context into `WebShutdown`, and from `LocalBackend.Shutdown` provide it with a five second timeout context, similar to what we do for logout.

Reply:
only worry is long wait times while holding the lock, so if we can release the lock and then continue shutting down i think its fine.

Reply:
It could still block LocalBackend.Shutdown from completing the rest of its shutdown work.

Reply:
yeah, agreed we would need to handle that too though thats less worrisome because typically the process is about to exit anyways. We would only see that impact in tests, but if shutdown blocks for a long time while holding a lock during the normal course of the process then we can have all sorts of weird things happen :/